sound/soc/sof/intel/probe.c

```diff
@@ -0,0 +1,199 @@
+// SPDX-License-Identifier: (GPL-2.0 OR BSD-3-Clause)
+//
+// This file is provided under a dual BSD/GPLv2 license.  When using or
+// redistributing this file, you may do so under either license.
+//
+// Copyright(c) 2019 Intel Corporation. All rights reserved.
+//
+// Author: Cezary Rojewski <cezary.rojewski@intel.com>
+//
+
+#include <sound/hdaudio_ext.h>
+#include "../sof-priv.h"
+#include "probe.h"
+
+int sof_ipc_probe_init(struct snd_sof_dev *sdev,
+		size_t buffer_size)
+{
+	struct sof_ipc_probe_dma_set_params *msg;
+	struct sof_ipc_reply reply;
+	unsigned int stream_tag = INVALID_NODE_ID;
+	size_t size;
+	int ret;
+
+	size = struct_size(msg, dma, 1);
+	msg = kzalloc(size, GFP_KERNEL);
+	if (!msg)
+		return -ENOMEM;
+
+	if (sdev->extractor)
+		stream_tag = sdev->extractor->hstream.stream_tag - 1;
+
+	msg->hdr.size = size;
+	msg->hdr.cmd = SOF_IPC_GLB_PROBE | SOF_IPC_PROBE_INIT;
+	msg->dma[0].stream_tag = stream_tag;
+	msg->dma[0].dma_buffer_size = buffer_size;
+
+	ret = sof_ipc_tx_message(sdev->ipc, msg->hdr.cmd, msg, msg->hdr.size,
+			&reply, sizeof(reply));
+	kfree(msg);
+	return ret;
+}
+
+int sof_ipc_probe_deinit(struct snd_sof_dev *sdev)
+{
+	struct sof_ipc_cmd_hdr msg;
+	struct sof_ipc_reply reply;
+
+	msg.size = sizeof(msg);
+	msg.cmd = SOF_IPC_GLB_PROBE | SOF_IPC_PROBE_DEINIT;
+
+	return sof_ipc_tx_message(sdev->ipc, msg.cmd, &msg, msg.size,
+			&reply, sizeof(reply));
+}
+
+static int sof_ipc_probe_get(struct snd_sof_dev *sdev, unsigned int cmd,
+		void **params, size_t *num_params)
+{
+	struct sof_ipc_probe_get_params msg = {0};
+	struct sof_ipc_probe_get_params *reply;
+	size_t size;
+	int ret;
+
+	*params = NULL;
+	*num_params = 0;
+
+	reply = kzalloc(SOF_IPC_MSG_MAX_SIZE, GFP_KERNEL);
+	if (!reply)
+		return -ENOMEM;
+	msg.rhdr.hdr.size = sizeof(msg);
+	msg.rhdr.hdr.cmd = SOF_IPC_GLB_PROBE | cmd;
+
+	ret = sof_ipc_tx_message(sdev->ipc, msg.rhdr.hdr.cmd, &msg,
+			msg.rhdr.hdr.size, reply, SOF_IPC_MSG_MAX_SIZE);
+	if (ret < 0 || reply->rhdr.error < 0)
+		goto exit;
+
+	size = reply->rhdr.hdr.size -= sizeof(reply);
+	if (size <= 0)
+		goto exit;
+
+	*params = kzalloc(size, GFP_KERNEL);
+	if (!*params) {
+		ret = -ENOMEM;
+		goto exit;
+	}
+	memcpy(*params, &reply->dma[0], size);
+	*num_params = size;
+
+exit:
+	kfree(reply);
+	return ret;
+}
+
+int sof_ipc_probe_get_dma(struct snd_sof_dev *sdev,
+		struct sof_probe_dma **dma, size_t *num_dma)
+{
+	int ret;
+
+	ret = sof_ipc_probe_get(sdev, SOF_IPC_PROBE_DMA_GET,
+			(void **)dma, num_dma);
+	*num_dma = *num_dma / sizeof(**dma);
```

Comment:
looks like a bad definition of that sof_icp_probe_get() does if you need to patch the values returned?

Reply:
Decided to update the _probe_get definition to:
static int sof_ipc_probe_get(struct snd_sof_dev *sdev, unsigned int cmd,
		void **params, size_t *bytes)

Difference:
s/size_t *num_params/size_t *bytes/
Should prove more readable.

In regard to your question: Arthur implemented _GET as single struct where both flexible arrays (dma AND probe points) are located together - within an union. And so, I've decided to have a single, common _get method to reduce code size. As _GET does not know what type of payload it is actually carrying - whether it's dma or probe_point list - one cannot assume the divisor (struct size). Additional division is a small cost for the overall gain.

Reply:
not able to understand the reply, so will wait for the v2.